### DISCUSSION TITLE
What determines the source IP of outbound packets?

### DISCUSSION THREAD

**Discussion by TomHsiung** [strongswan/strongswan#1680] [General]
Hey, guys

I want to know more detail about my experience today.

My gateway host was included in the site-to-site IPsec tunnel. It worked fine until today when I discovered that its outbound traffic was not entering the tunnel. `A.A.A.A` is my public IP address derived from my PPPoE.

> ping -I ppp0 -s 234 -D 10.5.96.4
> PING 10.5.96.4 (10.5.96.4) from A.A.A.A ppp0: 234(262) bytes of data.
> ^C
> --- 10.5.96.4 ping statistics ---
> 8 packets transmitted, 0 received, 100% packet loss, time 7171ms

The host use the public IP address as the source IP of its outbound packets. As a result, these traffic will not enter the site-to-site tunnel based on configured IPsec policies.

However, after I restart the IPsec via `ipsec restart`, I got this. The host use `10.0.0.1` as its source IP address. Could you please tell me the factors which affect how the host selects the source IP for its outbound packets? Thanks.

> ping -I ppp0 -s 234 -D 10.5.96.4
> ping: Warning: source address might be selected on device other than: ppp0
> PING 10.5.96.4 (10.5.96.4) from 10.0.0.1 ppp0: 234(262) bytes of data.
> [1683394925.314207] 242 bytes from 10.5.96.4: icmp_seq=1 ttl=63 time=205 ms
> [1683394926.310091] 242 bytes from 10.5.96.4: icmp_seq=2 ttl=63 time=200 ms
> [1683394927.312010] 242 bytes from 10.5.96.4: icmp_seq=3 ttl=63 time=201 ms
> [1683394928.320571] 242 bytes from 10.5.96.4: icmp_seq=4 ttl=63 time=208 ms
> ^C
> --- 10.5.96.4 ping statistics ---
> 4 packets transmitted, 4 received, 0% packet loss, time 3003ms
> rtt min/avg/max/mdev = 200.025/203.653/208.472/3.430 ms

**Comment by tobiasbrunner** [comment]
strongSwan installs a route in table 220 that sets the preferred source address to one in the local traffic selector (see `ip route list table 220`). If the interface goes down, the route gets removed, however, it should get reinstalled if the interface comes back up. Maybe that fails on your host for some reason (check the log, maybe with the [log level](https://docs.strongswan.org/docs/5.9/config/logging.html) for _knl_ increased to 2).

**Reply by TomHsiung** [reply]
> Maybe that fails on your host for some reason

I have confirmed that it happened when I plugged out the cable which connects the ethernet (`10.0.0.1`) to the downstream switch. The reason I plugged out the cable is that I intended to deploy a passive sniffer between `10.0.0.1` and the switch. Is there a route priority in the kernel?

In addition, the gateway runs strongSwan and the later listens on both the ppp0 (`public IP`) and the `10.0.0.1`.

**Reply by tobiasbrunner** [reply]
As I said, check the routes and the log. In regards to the priority, routes have metrics and strongSwan installs a routing rule so table 220 is preferred over the main routing table (see `ip rule`).

**Reply by TomHsiung** [reply]
Thanks for your suggestion. Forgot to illustrate, that, the passive network tap would face the network to switch from 1000baseT to 100baseTX. I will post the routes before and after the unplugging-and-replugging behavior later.

**Comment by TomHsiung** [comment]
And, I read some webpages about source IP address selection. It seems that the outbound packets of the host on which strongSwan runs will select their source IP address based on content in the routing table. So, strongSwan added the `route table 220` into the kernel, which makes the host choose the interface and the IP address related to it to send out packets, no matter if the host itself is included in the tunnel. Correct?

I want to understand deeper. Both the PPPoE and internal private interface are on the same host and strongSwan is listening on both interfaces, why does not the host select the PPPoE interface to send out packets?

**Reply by tobiasbrunner** [reply]
The route is installed to select a source IP from the local traffic selector so locally generated traffic matches the IPsec policies. The interface is the outbound interface (the one over which the peer's public IP is reached). Increase the log level for _knl_ to see details on this route installation.

Which interface is selected for IKE/ESP packets depends on the other routes installed on your system.

**Reply by TomHsiung** [reply]
Don't fully understand this process. Would you suggest some good resources for the explanation about this? Thank you.

**Reply by tobiasbrunner** [reply]
Which process?

**Reply by TomHsiung** [reply]
Hmmm, I don't quite understand the packet flow process. I mean what is the relationship between the route table 220 and the iptable (xfrm-lookup, xfrm-encode). It looks like the route table 220 and the iptable work together to handle those packets matching the IPsec policy. But I don't know how.

**Reply by tobiasbrunner** [reply]
The routes in table 220 are necessary to make the kernel select a specific source IP for locally generated traffic addressed to the remote subnet(s) so that it will match an outbound IPsec policy. Without them the kernel would probably select the public IP as source address (via the default route), which might not be part of the local traffic selectors (i.e. there wouldn't be a policy matching such traffic and it wouldn't get encrypted).

**Reply by TomHsiung** [reply]
So, for forwarded packets (from other hosts in the subnets rather than the strongSwan server itself), they do not need the routing table 220. Hmmm, when these packets enter into the iptable flow they already have source IP addresses that match the policy.

**Reply by tobiasbrunner** [reply]
Yes, the source IP for forwarded traffic won't change (unless there is a NAT). And the routes in table 220 are only needed for such traffic if you don't have a route for the remote subnets in the main routing table (but that's unlikely as there is probably a default route anyway).
